### ISSUE TITLE
`pl-richtext-editor` swallows extra line breaks

### ISSUE THREAD

**Issue by firasm** [PrairieLearn/PrairieLearn#10026]
Definitely unrelated to the recent changes to pl-richtext-editor (#10017, #10012), it seems extra linebreaks are removed from the input.

The culprit might be this line based on the `DOMPurify()` package, but not 100% sure:

https://github.com/PrairieLearn/PrairieLearn/blob/161f7518acaf8aaf0ded29d645d9337e4ad7ce90/apps/prairielearn/elements/pl-rich-text-editor/pl-rich-text-editor.js#L93

Is this configurable to pass line breaks through ? 

Grading essays without (>1) line breaks will be rough to do 😬.

![pl_richtext_space](https://github.com/PrairieLearn/PrairieLearn/assets/2507459/8a69e81a-755f-4c23-b437-8ecdcfdcffdf)

**Comment by jonatanschroeder** [comment]
I don't see this problem locally. Which OS/browser are you using? Do you see the same behaviour in other browsers?

**Comment by firasm** [comment]
I tried it on Safari and Firefox, able to reproduce on both. I'm on macOS. I'll try and couple of other browsers and report back. 

Also, I found this while I was troubleshooting: https://github.com/cure53/DOMPurify/issues/830

**Comment by firasm** [comment]
I was able to reproduce the missing line breaks in:

- Safari
- Chrome
- Firefox
- Vivaldi
- Arc

I even tried a barebones minimal example with no `server.py` and just this in `question.html`

```
<div class="card my-2">
<div class="card-header">Essay Quiz 10</div>

<div class="card-body">
    
<pl-question-panel>
<markdown>

Test Question

</markdown>
</pl-question-panel>

<pl-rich-text-editor file-name="essay10.md" format="markdown" quill-theme="snow" counter="character">
</pl-rich-text-editor>


</div>
</div>
```

**Comment by jonatanschroeder** [comment]
Does this happen with HTML format or only markdown?

**Comment by firasm** [comment]
I **cannot** reproduce the issue after changing `format="html"` !

**Comment by jonatanschroeder** [comment]
Ok, I'm guessing this is getting lost in the conversation to or from Markdown.

Related: #6480.

**Comment by jonatanschroeder** [comment]
I was able to reproduce this locally if the format is Markdown. It seems to be related to showdown (I believe https://github.com/showdownjs/showdown/issues/801 is related). In essence Markdown itself [treats any sequence of 2 or more line breaks as a paragraph break](https://spec.commonmark.org/0.17/#example-131), so converting a blank paragraph in HTML as a sequence of line breaks seems to be an incorrect interpretation by showdown. I haven't had a chance to test other libraries that convert HTML to Markdown, but this would likely require something like adding an artificial space into the blank paragraph.

Anyway, my suggestion is that, unless you have a strong reason to use `format="markdown"`, you stick to the default HTML format. There are other issues with markdown that can popup in a few cases (e.g., #5986), and we'll eventually be able to handle that, but they'd likely be handled in the context of #6480.

**Comment by echuber2** [comment]
We could add some option for the Markdown mode that would preserve extra lines in a Markdown-compatible way. For example, if you want students to use single empty lines in some places and double empty lines in other places, the double empty line section can be preserved like this:

```
x

<br>
y
```

Similarly, three empty lines can be preserved like this:

```
x

<br>
<br>
y
```

The normal empty line ensures that a paragraph break happens, and the extra `<br>` add the remaining line breaks. Here's an alternative way to encode three empty lines, with the `<br>` tags associated as trailing elements instead of head elements:

```
x<br><br><br>

y
```

I think a benefit of using Markdown features to implement this workaround is that it should work in the future even if PL switches to a different Markdown library.

**Comment by jonatanschroeder** [comment]
> We could add some option for the Markdown mode that would preserve extra lines in a Markdown-compatible way.

I agree that this would be ideal, but I don't think it's worth spending time on trying to implement this in showdown, since other issues in it would likely lead us to a different library anyway. The issue in showdown is that `fromMarkdown` works relatively well, but `toMarkdown` has kinks like these (see the other issues listed above, which have been raised with showdown but with no response).
